drivers/flash/spi_nand.c

```diff
@@ -0,0 +1,1034 @@
+/*
+ * Copyright (c) 2022 Macronix International Co., Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT jedec_spi_nand
+
+#include <errno.h>
+#include <drivers/flash.h>
+#include <drivers/spi.h>
+#include <init.h>
+#include <string.h>
+#include <logging/log.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "spi_nand.h"
+#include "bch.h"
+
+LOG_MODULE_REGISTER(spi_nand, CONFIG_FLASH_LOG_LEVEL);
+
+/* Indicates that an access command includes bytes for the address.
+ * If not provided the opcode is not followed by address bytes.
+ */
+#define NAND_ACCESS_ADDRESSED BIT(0)
+
+/* Indicates that addressed access uses a 8-bit address regardless of
+ * spi_nand_data::flag_8bit_addr.
+ */
+#define NAND_ACCESS_8BIT_ADDR BIT(1)
+
+/* Indicates that addressed access uses a 16-bit address regardless of
+ * spi_nand_data::flag_16bit_addr.
+ */
+#define NAND_ACCESS_16BIT_ADDR BIT(2)
+
+/* Indicates that addressed access uses a 24-bit address regardless of
+ * spi_nand_data::flag_32bit_addr.
+ */
+#define NAND_ACCESS_24BIT_ADDR BIT(3)
+
+/* Indicates that addressed access uses a 32-bit address regardless of
+ * spi_nand_data::flag_32bit_addr.
+ */
+#define NAND_ACCESS_32BIT_ADDR BIT(4)
+
+/* Indicates that an access command is performing a write.  If not
+ * provided access is a read.
+ */
+#define NAND_ACCESS_WRITE BIT(5)
+
+#define NAND_ACCESS_DUMMY BIT(6)
+
+/* Build-time data associated with the device. */
+struct spi_nand_config {
+	/* Devicetree SPI configuration */
+	struct spi_dt_spec spi;
+
+	uint8_t id[2];
+	uint8_t support_conti_read;
+};
+
+/**
+ * struct spi_nand_data - Structure for defining the SPI NAND access
+ * @sem: The semaphore to access to the flash
+ */
+struct spi_nand_data {
+	struct k_sem sem;
+
+	/* Number of bytes per page */
+	uint16_t page_size;
+
+	/* Number of oob bytes per page */
+	uint8_t oob_size;
+
+	/* Number of pages per block */
+	uint16_t page_num;
+
+	/* Number of bytes per block */
+	uint32_t block_size;
+
+	/* Number of blocks per chip */
+	uint16_t block_num;
+
+	/* Size of flash, in bytes */
+	uint32_t flash_size;
+
+	uint8_t page_shift;
+
+	uint8_t block_shift;
+
+	bool continuous_read;
+
+	bool read_recovery;
+
+	uint8_t ecc_bits;
+
+	uint8_t ecc_bytes;
+	uint8_t ecc_steps;
+	uint8_t ecc_layout_pos;
+	uint32_t  ecc_size;
+	uint8_t *ecc_calc;
+	uint8_t *ecc_code;
+	uint8_t *page_buf;
+
+	struct nand_bch_control {
+		struct bch_code      *bch;
+		unsigned int         *errloc;
+		unsigned char        *eccmask;
+	}nbc;
+};
+
+static const struct flash_parameters flash_nand_parameters = {
+	.write_block_size = SPI_NAND_SUB_PAGE_SIZE,
+	.erase_value = 0xff,
+};
+
+void bch_calculate_ecc(const struct device *dev, 
+		unsigned char *buf, unsigned char *code)
+{
+	struct spi_nand_data *data = dev->data;
+
+	memset(code, 0, data->ecc_bytes);
+
+	bch_encode(data->nbc.bch, buf, (unsigned int *)code);
+}
+
+void bch_correct_data(const struct device *dev, 
```

Comment:
Does this need to have a global scope?

Reply:
Because this function is defined before calling its function, I believe there is no need for a global declaration